File: tmdb-core/src/main/java/dev/reuss/tmdb/domain/collection/DefaultCollectionService.java
package dev.reuss.tmdb.domain.collection;

import dev.reuss.tmdb.common.image.CollectionImages;
import dev.reuss.tmdb.core.http.TmdbHttpClient;
import dev.reuss.tmdb.core.http.TmdbRequest;
import dev.reuss.tmdb.domain.collection.model.CollectionDetails;
import dev.reuss.tmdb.domain.collection.model.CollectionTranslations;
import dev.reuss.tmdb.domain.collection.query.CollectionDetailsQuery;
import dev.reuss.tmdb.query.ImageQuery;
import dev.reuss.tmdb.value.id.CollectionId;
import dev.reuss.tmdb.value.language.Language;

import java.util.Objects;

/**
 * Default collection service implementation.
 */
public final class DefaultCollectionService implements CollectionService {

    private final TmdbHttpClient httpClient;

    public DefaultCollectionService(TmdbHttpClient httpClient) {
        this.httpClient = Objects.requireNonNull(httpClient, "HTTP client must not be null");
    }

    @Override
    public CollectionDetails details(CollectionId collectionId) {
        return details(collectionId, CollectionDetailsQuery.empty());
    }

    @Override
    public CollectionDetails details(CollectionId collectionId, Language language) {
        return details(collectionId, CollectionDetailsQuery.of(language));
    }

    @Override
    public CollectionDetails details(CollectionId collectionId, CollectionDetailsQuery query) {
        Objects.requireNonNull(collectionId, "Collection id must not be null");
        Objects.requireNonNull(query, "Collection details query must not be null");

        return httpClient.get(
                TmdbRequest.get(CollectionPaths.details(collectionId), query.toQueryParams()),
                CollectionDetails.class
        );
    }

    @Override
    public CollectionImages images(CollectionId collectionId) {
        return images(collectionId, ImageQuery.none());
    }

    @Override
    public CollectionImages images(CollectionId collectionId, ImageQuery query) {
        Objects.requireNonNull(collectionId, "Collection id must not be null");
        Objects.requireNonNull(query, "Image query must not be null");

        return httpClient.get(
                TmdbRequest.get(CollectionPaths.images(collectionId), query.toQueryParams()),
                CollectionImages.class
        );
    }

    @Override
    public CollectionTranslations translations(CollectionId collectionId) {
        Objects.requireNonNull(collectionId, "Collection id must not be null");

        return httpClient.get(
                TmdbRequest.get(CollectionPaths.translations(collectionId)),
                CollectionTranslations.class
        );
    }
}

File: tmdb-core/src/main/kotlin/dev/reuss/tmdb/domain/collection/DefaultCollectionService.kt
package dev.reuss.tmdb.domain.collection

import dev.reuss.tmdb.common.image.CollectionImages
import dev.reuss.tmdb.core.http.TmdbHttpClient
import dev.reuss.tmdb.core.http.TmdbRequest
import dev.reuss.tmdb.domain.collection.model.CollectionDetails
import dev.reuss.tmdb.domain.collection.model.CollectionTranslations
import dev.reuss.tmdb.domain.collection.query.CollectionDetailsQuery
import dev.reuss.tmdb.query.ImageQuery
import dev.reuss.tmdb.value.id.CollectionId
import dev.reuss.tmdb.value.language.Language

/**
 * Default collection service implementation.
 */
internal class DefaultCollectionService(
    private val httpClient: TmdbHttpClient,
) : CollectionService {
    override fun details(collectionId: CollectionId): CollectionDetails =
        details(collectionId, CollectionDetailsQuery.empty())

    override fun details(
        collectionId: CollectionId,
        language: Language,
    ): CollectionDetails = details(collectionId, CollectionDetailsQuery.of(language))

    override fun details(
        collectionId: CollectionId,
        query: CollectionDetailsQuery,
    ): CollectionDetails =
        httpClient.get(
            TmdbRequest.get(
                CollectionPaths.details(collectionId),
                query.toQueryParams(),
            ),
            CollectionDetails::class.java,
        )

    override fun images(collectionId: CollectionId): CollectionImages = images(collectionId, ImageQuery.none())

    override fun images(
        collectionId: CollectionId,
        query: ImageQuery,
    ): CollectionImages =
        httpClient.get(
            TmdbRequest.get(
                CollectionPaths.images(collectionId),
                query.toQueryParams(),
            ),
            CollectionImages::class.java,
        )

    override fun translations(collectionId: CollectionId): CollectionTranslations =
        httpClient.get(
            TmdbRequest.get(CollectionPaths.translations(collectionId)),
            CollectionTranslations::class.java,
        )
}
